### ISSUE TITLE
REINFORCE++ drops outcome rewards before multi-turn observation spans

### ISSUE THREAD

**Issue by ai-yang** [verl-project/verl#7278]
## System Info

- verl main commit: `2781c1e405fa80871a838bfdf9495b412df19c35`
- Linux x86_64, kernel 5.15
- Python 3.12.13
- PyTorch 2.13.0+cpu
- tensordict 0.10.0
- NumPy 2.5.1
- CUDA was unavailable; this reproducer exercises the device-independent advantage calculation on CPU.

## Information

- [x] The official REINFORCE++ advantage implementation
- [x] A minimal standalone reproduction

## Tasks

- [x] Multi-turn RL trajectories containing tool observations

## Reproduction

From a checkout of the commit above:

```bash
.venv/bin/python - <<'PY'
from types import SimpleNamespace
import torch
from verl.trainer.ppo.core_algos import compute_reinforce_plus_plus_outcome_advantage

config = SimpleNamespace(gamma=1.0)
compact_mask = torch.ones(1, 4)
compact_rewards = torch.tensor([[0.0, 0.0, 0.0, 1.0]])
expanded_mask = torch.tensor([[1.0, 1.0, 0.0, 0.0, 1.0, 1.0]])
expanded_rewards = torch.tensor([[0.0, 0.0, 0.0, 0.0, 0.0, 1.0]])

for run in range(1, 4):
    _, compact = compute_reinforce_plus_plus_outcome_advantage(
        compact_rewards, compact_mask, config=config
    )
    _, expanded = compute_reinforce_plus_plus_outcome_advantage(
        expanded_rewards, expanded_mask, config=config
    )
    print(
        f"run={run}",
        f"compact={compact.flatten().tolist()}",
        f"with_observation={expanded[expanded_mask.bool()].tolist()}",
    )
PY
```

This inserts two zero-masked observation tokens between the same four assistant tokens. With `gamma=1`, the valid-token returns should be invariant to that insertion. The baseline instead prints this result on all three runs:

```text
run=1 compact=[1.0, 1.0, 1.0, 1.0] with_observation=[0.0, 0.0, 1.0, 1.0]
run=2 compact=[1.0, 1.0, 1.0, 1.0] with_observation=[0.0, 0.0, 1.0, 1.0]
run=3 compact=[1.0, 1.0, 1.0, 1.0] with_observation=[0.0, 0.0, 1.0, 1.0]
```

`AgentLoopManager.generate_sequences` documents `response_mask=0` for both tool observations and trailing padding. In `compute_reinforce_plus_plus_outcome_advantage`, the reverse scan currently multiplies `running_return` by that mask at every position. An internal observation therefore resets the return as if the episode ended, preventing the final outcome reward from reaching earlier assistant turns.

The existing multi-turn GAE recurrence already treats observation tokens as skipped positions: they neither terminate the trajectory nor consume a discount step.

## Expected behavior

Masked tool-observation positions should be skipped while computing REINFORCE++ returns. Trailing padding should remain zero, and valid assistant-token returns should match the same action sequence without inserted observations. For the reproduction above, both valid-token return vectors should be `[1.0, 1.0, 1.0, 1.0]`.
